upstairs/src/client.rs

```diff
@@ -0,0 +1,3133 @@
+    /// Returns an `IOop` with modified dependencies
+    pub(crate) fn in_progress(
+        &mut self,
+        job: &mut DownstairsIO,
+        repair_min_id: Option<JobId>,
+    ) -> IOop {
+        let old_state = self.set_job_state(job, IOState::InProgress);
+        assert_eq!(old_state, IOState::New);
+
+        let mut out = job.work.clone();
+        if self.dependencies_need_cleanup() {
+            match &mut out {
+                IOop::Write { dependencies, .. }
+                | IOop::WriteUnwritten { dependencies, .. }
+                | IOop::Flush { dependencies, .. }
+                | IOop::Read { dependencies, .. }
+                | IOop::ExtentClose { dependencies, .. }
+                | IOop::ExtentFlushClose { dependencies, .. }
+                | IOop::ExtentLiveRepair { dependencies, .. }
+                | IOop::ExtentLiveReopen { dependencies, .. }
+                | IOop::ExtentLiveNoOp { dependencies } => {
+                    self.remove_dep_if_live_repair(
+                        dependencies,
+                        job.ds_id,
+                        repair_min_id.expect("must have repair_min_id"),
+                    );
+                }
+            }
+        }
+        // If our downstairs is under repair, then include any extent limit sent
+        // in the IOop; otherwise, clear it out
+        if let IOop::Flush { extent_limit, .. } = &mut out {
+            if !matches!(self.state, DsState::LiveRepair) {
+                *extent_limit = None;
+            }
+        }
+        out
+    }
+
+    pub(crate) fn replay_job(&mut self, job: &mut DownstairsIO) {
+        /*
+         * If the job is InProgress or New, then we can just go back
+         * to New and no extra work is required.
+         * If it's Done, then by definition it has been acked; assert that here
+         * to double-check.
+         */
+        if IOState::Done == job.state[self.client_id] {
+            assert!(job.acked);
+        }
+        let old_state = self.set_job_state(job, IOState::New);
+        job.replay = true;
+        if old_state != IOState::New {
+            self.requeue_one(job.ds_id);
+        }
+    }
+
+    /// Sets this job as skipped and moves it to `skipped_jobs`
+    pub(crate) fn skip_job(&mut self, job: &mut DownstairsIO) {
+        self.set_job_state(job, IOState::Skipped);
+        self.skipped_jobs.insert(job.ds_id);
+    }
+
+    /// Returns true if it's possible that we need to clean job dependencies
+    pub(crate) fn dependencies_need_cleanup(&self) -> bool {
+        matches!(self.state, DsState::LiveRepair)
+            && !self.skipped_jobs.is_empty()
+    }
+
+    /// Sets our state to `DsState::Repair`
+    ///
+    /// # Panics
+    /// If the previous state is not `DsState::WaitQuorum`
+    pub(crate) fn begin_reconcile(&mut self) {
+        /*
+         * We transition all the downstairs to needing repair here
+         * while we have the downstairs lock.  This will insure that
```

Comment:
Good catch, done in 9c95e2c68868e34f32c2b21bb10b359b8d3cfcff (I also removed the comment about the panic, because we list it as a precondition in the docstring)